include/circt/Support/CPPCDE.h

```diff
@@ -0,0 +1,398 @@
+  // be done at CIRCT verification time.
+  // For this initial version, just use the latter.
+
+  virtual CDEValue operator+(CDEValue other) {
+    return execBinOp<typename TOps::Add>(other);
+  }
+  virtual CDEValue operator-(CDEValue other) {
+    return execBinOp<typename TOps::Sub>(other);
+  }
+  virtual CDEValue operator*(CDEValue other) {
+    return execBinOp<typename TOps::Mul>(other);
+  }
+  virtual CDEValue operator<<(CDEValue other) {
+    return execBinOp<typename TOps::Shl>(other);
+  }
+  virtual CDEValue operator&(CDEValue other) {
+    return execBinOp<typename TOps::And>(other);
+  }
+  virtual CDEValue operator|(CDEValue other) {
+    return execBinOp<typename TOps::Or>(other);
+  }
+  virtual CDEValue operator^(CDEValue other) {
+    return execBinOp<typename TOps::XOr>(other);
+  }
+  virtual CDEValue operator~() {
+    return CDEValue(
+        this->ctx,
+        comb::createOrFoldNot(this->get().getLoc(), this->get(), this->ctx->b));
+  }
+  virtual CDEValue shrs(CDEValue other) {
+    return execBinOp<typename TOps::ShrS>(other);
+  }
+  virtual CDEValue shru(CDEValue other) {
+    return execBinOp<typename TOps::ShrU>(other);
+  }
+  virtual CDEValue divu(CDEValue other) {
+    return execBinOp<typename TOps::DivU>(other);
+  }
+  virtual CDEValue divs(CDEValue other) {
+    return execBinOp<typename TOps::DivS>(other);
+  }
+  virtual CDEValue modu(CDEValue other) {
+    return execBinOp<typename TOps::ModU>(other);
+  }
+  virtual CDEValue mods(CDEValue other) {
+    return execBinOp<typename TOps::ModS>(other);
+  }
+  virtual CDEValue concat(CDEValue other) {
+    return execBinOp<typename TOps::Concat>(other);
+  }
+
+protected:
+  CDEContext *ctx = nullptr;
+  std::variant<Value, Backedge *> valueOrBackedge;
+};
+
+class ESICDEValue : public CDEValueImpl<CDEOperatorSelector, ESICDEValue> {
+public:
+  using CDEValueImpl::CDEValueImpl;
+
+  std::pair<ESICDEValue, ESICDEValue> wrap(ESICDEValue valid) {
+    auto wrapOp = ctx->b.create<esi::WrapValidReadyOp>(
+        get().getLoc(), this->get(), valid.get());
+    return std::make_pair(ESICDEValue(ctx, wrapOp.getResult(0)),
+                          ESICDEValue(ctx, wrapOp.getResult(1)));
+  }
+
+  std::pair<ESICDEValue, ESICDEValue> unwrap(ESICDEValue ready) {
+    auto unwrapOp = this->ctx->b.template create<esi::UnwrapValidReadyOp>(
+        this->get().getLoc(), this->get(), ready.get());
+    return std::make_pair(ESICDEValue(this->ctx, unwrapOp.getResult(0)),
+                          ESICDEValue(this->ctx, unwrapOp.getResult(1)));
+  }
+};
+
+struct DefaultCDEValue
+    : public CDEValueImpl<CDEOperatorSelector, DefaultCDEValue> {
+  using CDEValueImpl::CDEValueImpl;
+};
+
+// A class for providing backedge-based access to the in- and output ports of
+// a module.
+template <typename CDEValue>
+class CDEPorts {
+
+public:
+  CDEPorts(CDEContext &ctx, hw::HWModuleOp module) : ctx(ctx) {
```

Comment:
Is it possible to use HWModuleLike here?

Reply:
I was just looking at adding a pass that would insert output ports. I was trying to use the HWMutableModuleLike interface, but realized it doesn't support mutating the OutputOp. However, the HW module does: https://github.com/llvm/circt/blob/main/include/circt/Dialect/HW/HWStructure.td#L102. I was thinking we could make that accessible through the interface. Would that work here @mortbopet ?

Reply:
> Would that work here @mortbopet ?

Sounds like it would!

Reply:
alright, i'm opening an issue so we can track it: https://github.com/llvm/circt/issues/3791

Reply:
I haven't looked too deeply at what's going on here, but for my use case, it was enough to expose appendOutputs in the HWMutableModuleLike interface: https://github.com/llvm/circt/pull/3800.